walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,101 @@
+    MeetingRequest requestEveryone = new MeetingRequest(allAttendees, request.getDuration());
+    Collection<TimeRange> queryEveryone = queryGivenAttendees(events, requestEveryone);
+		
+    if (!queryEveryone.isEmpty()) {
+      return queryEveryone;
+    }
+    else {
+      return queryGivenAttendees(events, request);
+    }
+  }
+
+  /** Returns the possible meeting times given required attendees */
+  public Collection<TimeRange> queryGivenAttendees(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> busy = new ArrayList<>();
+		
+    // Determine times where meeting cannot be scheduled
+    for (Event event : events) { 
+      if (overlappingAttendees(event.getAttendees(), request.getAttendees())) {
+        busy.add(event.getWhen());
+      }
+    }
+
+    // Sort these times by start time
+    Collections.sort(busy, TimeRange.ORDER_BY_START);
+
+    // Simplify the times to avoid overlap
+    ArrayList<TimeRange> mergedBusy = merge(busy); 
+
+    // Get the complement 
+    Collection<TimeRange> free = new ArrayList<>();
+    if (mergedBusy.size() > 0) { 
+      // Add time at start of day if possible
+      if (mergedBusy.get(0).start() - TimeRange.START_OF_DAY >= request.getDuration()) {
+        free.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, mergedBusy.get(0).start(), false));
+      }
+      // Fill in the time between unavailable times
+      for (int i = 0; i < mergedBusy.size() - 1; i++) {
+        if (mergedBusy.get(i + 1).start() - mergedBusy.get(i).end() >= request.getDuration()) {
+          free.add(TimeRange.fromStartEnd(mergedBusy.get(i).end(), mergedBusy.get(i + 1).start(), false));
+        }
+      }
+      // Add time at end of day if possible
+      if (TimeRange.END_OF_DAY - mergedBusy.get(mergedBusy.size() - 1).end() >= request.getDuration()) {
+        free.add(TimeRange.fromStartEnd(mergedBusy.get(mergedBusy.size() - 1).end(), TimeRange.END_OF_DAY, true));
+      }
+    }
+    // Add the whole day if possible
+    else if (TimeRange.WHOLE_DAY.duration() >= request.getDuration() && request.getAttendees().size() > 0) {
+      free.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    }
+    return free;
+  }
+
+  /** Eliminates the overlapping times in a sorted-by-starttime list of TimeRange objects*/
+  private ArrayList<TimeRange> merge(ArrayList<TimeRange> list) {
+    // invariant: merged contains non-overlapping TimeRange objects
+    ArrayList<TimeRange> merged = new ArrayList<TimeRange>();
+
+    for (int i = 0; i < list.size(); i++) {
+      int lastMergedIndex = merged.size() - 1;
+      // Empty or there is no overlap, so add in the TimeRange
+      if (merged.size() == 0 || !merged.get(lastMergedIndex).overlaps(list.get(i))) {
+        merged.add(list.get(i));
+      }
+      // There is uncontained overlap, so replace last element
+      else if (!merged.get(lastMergedIndex).contains(list.get(i))) {
```

Comment:
nit: This `else if` should be in the same like as the `}` above.

Reply:
Should this be on line line 95 here?

Reply:
Looks like line 78 to me, but in general if-else statements should read:

```
if (somecondtion) {
 // Do something
} else if (someothercondition) {
 // Do something else
} else {
 // Do something else else
}
```

Reply:
Thanks, I fixed this for all if-else statements in the document